File: ios/masks/DatetimeMaskEngine.swift
import NitroModules

struct DatetimeMaskEngine: MaskEngineProtocol {
  private let compiled: CompiledMask

  init(format: String) {
    let pattern = DatetimeMaskEngine.formatToPattern(format)
    self.compiled = MaskEngine.compile(mask: pattern)
  }

  func apply(input: String) -> (masked: String, raw: String) {
    return MaskEngine.apply(input: input, compiled: compiled)
  }

  var wantsTrailingCursor: Bool { false }
  var expandedMask: String? { compiled.expandedMask }

  // MARK: - Format → Pattern conversion

  private static let tokens: [(token: String, pattern: String)] = [
    ("YYYY", "9999"),
    ("MM", "[1-12]"),
    ("DD", "[1-31]"),
    ("HH", "[0-23]"),
    ("hh", "[1-12]"),
    ("mm", "[0-59]"),
    ("ss", "[0-59]"),
  ]

  private static func formatToPattern(_ format: String) -> String {
    var result = ""
    var i = format.startIndex

    while i < format.endIndex {
      if let match = tokens.first(where: { format[i...].hasPrefix($0.token) }) {
        result += match.pattern
        i = format.index(i, offsetBy: match.token.count)
      } else {
        result.append(format[i])
        i = format.index(after: i)
      }
    }

    return result
  }
}

File: android/src/main/java/com/nitroinputmask/masks/DatetimeMaskEngine.kt
package com.nitroinputmask.masks

import com.nitroinputmask.engine.CompiledMask
import com.nitroinputmask.engine.MaskEngine
import com.nitroinputmask.engine.MaskEngineProtocol

internal class DatetimeMaskEngine(format: String) : MaskEngineProtocol {
  private val compiled: CompiledMask = MaskEngine.compile(formatToPattern(format))

  override fun apply(input: String): Pair<String, String> {
    return MaskEngine.apply(input, compiled)
  }

  override val wantsTrailingCursor: Boolean = false
  override val expandedMask: String? = compiled.expandedMask

  companion object {
    private fun formatToPattern(format: String): String {
      val sb = StringBuilder()
      var i = 0

      while (i < format.length) {
        when {
          format.startsWith("YYYY", i) -> { sb.append("9999");    i += 4 }
          format.startsWith("MM",   i) -> { sb.append("[1-12]");  i += 2 }
          format.startsWith("DD",   i) -> { sb.append("[1-31]");  i += 2 }
          format.startsWith("HH",   i) -> { sb.append("[0-23]");  i += 2 }
          format.startsWith("hh",   i) -> { sb.append("[1-12]");  i += 2 }
          format.startsWith("mm",   i) -> { sb.append("[0-59]");  i += 2 }
          format.startsWith("ss",   i) -> { sb.append("[0-59]");  i += 2 }
          else -> { sb.append(format[i]); i++ }
        }
      }

      return sb.toString()
    }
  }
}
